packages/fxa-settings/src/lib/hooks/useValidate/index.tsx

```diff
@@ -0,0 +1,53 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import { useEffect, useMemo, useState } from 'react';
+import { validate, ValidationError } from 'class-validator';
+import { ReachRouterWindow } from '../../window';
+import { UrlQueryData } from '../../model-data';
+
+// ❌ TODO: improve types here. We may also want to tweak the shape of what this returns.
+export function useValidatedQueryParams<T extends Record<string, any>>(
+  Schema: new () => T
+) {
+  const [state, setState] = useState<{
+    queryParams: T | null;
+    queryParamErrors: Record<string, string> | null;
+  }>({ queryParams: null, queryParamErrors: null });
+
+  const urlQueryData = useMemo(
+    () => new UrlQueryData(new ReachRouterWindow()),
+    []
+  );
+
+  useEffect(() => {
+    const schema = new Schema();
+
+    for (let key in schema) {
```

Comment:
This doesn't iterate through the keys of an object, unless I don't understand what `schema` is at all.  Also would be nice to pass the data through the constructor instead.

Reply:
Hmm do you mean using `for ... of Object.keys(schema)` instead of `for ... in schema` to ensure enumerable properties?

Good point on passing to the constructor instead. I'm not super (:punhusky:) sure how to do that off the top of my head if I only want to iterate through the values from the schema (maybe I should just call it `model` instead) given `urlQueryData` is our data access store for query params; this doesn't seem right...

```
const data: Partial<T> = {};
for (let key of Object.keys(new Schema())) {
  const value = urlQueryData.get(key);
  if (value !== null) {
    data[key] = value;
  }
}
const schema = new Schema(data);
```

Reply:
> Hmm do you mean using for ... of Object.keys(schema) instead of for ... in schema

Oh I'm sorry.  I'm so used to for...of Object.keys that I didn't read that carefully.  I'll show myself the door.

Reply:
Heheh no sweat, using `Object.keys` is better here anyways. I'd like to take another look at this section regardless, types are awk and I'm not sure if getting back `queryParamErrors.email` for example is the nicest.